README.md
[![Last-changedate](https://img.shields.io/badge/last%20change-2017--06--29-brightgreen.svg)](https://github.com/rmflight/waitcopy/commits/master) [![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/waitcopy)](https://cran.r-project.org/package=waitcopy) [![Travis-CI Build Status](https://travis-ci.org/rmflight/waitcopy.svg?branch=master)](https://travis-ci.org/rmflight/waitcopy) [![Coverage Status](https://img.shields.io/codecov/c/github/rmflight/waitcopy/master.svg)](https://codecov.io/github/rmflight/waitcopy?branch=master) [![Licence](https://img.shields.io/github/license/mashape/apistatus.svg)](http://choosealicense.com/licenses/mit/) [![ORCiD](https://img.shields.io/badge/orcid-0000--0001--8141--7788-green.svg)](http://orcid.org/0000-0001-8141-7788)

waitcopy
========

Copy files during particular times of day and with metadata.

Useful Definitions
------------------

-   **filepath**: the full path to a given file, eg /home/user/file1.png
-   **basename**: the actual filename of a file path, eg file1.png in /home/user/file1.png
-   **dirname**: the directory portion of a filepath, eg /home/user in /home/user/file1.png

Description
-----------

Provides the function `wait_copy` that:

-   will only copy files during a set time interval
-   will wait a specific amount of time between file copies
-   creates a json file with some limited meta-data about the file
-   removes special characters from the file name

### Why??

Imagine someone you work with has a hard drive that you need data from, but that hard drive is only accessible via the network, mounted via SAMBA, and there are potentially duplicate files with the file or base name, files with the same file name that are different, and the file path provides some meta-information about the sample. In addition, the file names have odd characters in them (spaces, colons, etc) that make them a pain to work with from the command line on Linux, so you'd prefer if they weren't there.

The files are small, so even copying over the network is fast, but if you copy too many too quickly during the day, you'll get complaints about hitting this shared resource too often by the people who are local to it.

### The Solution

So ideally, you want to copy the files only during certain hours, wait a little bit between each copy operation, check for duplicates (via names and md5 hashing), strip the file name of special characters, and note where the file originated.

`waitcopy` provides these capabilities.

How it Works
------------

Given a file to copy, and a location to copy it to, does a few things:

-   strip special characters and spaces from the file name (if asked)
-   copy the file to a temp location
-   save the original path to the file it was being copied from
-   calculate the MD5 hash of the file
-   check master json data of MD5 hashes and file names
-   if MD5 is new, add the new file name, original file path, and MD5 to the master json file
-   if MD5 is not new, **add** the original file path to the matching file entry in the master json file.
-   if a matching file name is found but with a different MD5 hash, append the first 8 digits of the MD5 hash to the file name, and add it to the json data

Installation
------------

### Stable Version

``` r
# install.packages("devtools")
devtools::install_github("MoseleyBioinformaticsLab/waitcopy")
```

### Development Version

``` r
# install.packages("devtools")
devtools::install_github("rmflight/waitcopy")
```

Example Usage
-------------

### Worked Example

Lets imagine that we want to copy a set of files during a set time, and one of the files is duplicated (but we don't know that before we start).

``` r
library(waitcopy)
library(lubridate)
```

    ## 
    ## Attaching package: 'lubridate'

    ## The following object is masked from 'package:base':
    ## 
    ##     date

``` r
# files are in the extdata directory of waitcopy
testloc <- system.file("extdata", "set1", package = "waitcopy")

file_list <- dir(testloc, pattern = "raw", full.names = TRUE)
file_list
```

    ## [1] "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file1.raw"
    ## [2] "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file2.raw"
    ## [3] "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file3.raw"

We will setup a **temp** directory to copy them to:

``` r
temp_dir <- tempfile(pattern = "copyfiles-test-1")
dir.create(temp_dir)
dir(temp_dir)
```

    ## character(0)

And then lets set up to copy **20s** from now.

``` r
curr_time <- waitcopy:::get_now_in_local()
curr_today <- waitcopy:::get_today_in_local()

now_minus_today <- difftime(curr_time, curr_today, units = "s")
beg_time <- seconds(now_minus_today + 20)
end_time <- seconds(now_minus_today + 3600)

beg_time
```

    ## [1] "58278.6133356094S"

``` r
end_time
```

    ## [1] "61858.6133356094S"

And now let's copy! This is in the **near** future, so we will set the `wait_check` parameter to a low value of only 10 seconds, normally this is set to 30 minutes (1800 seconds), assuming that it is in the far future when you want to copy the files.

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"),
          start_time = beg_time, stop_time = end_time, wait_check = 10, pause_file = 0)
```

    ## Not allowed to copy yet, waiting! .... 2017-06-29 16:10:59

    ## Not allowed to copy yet, waiting! .... 2017-06-29 16:11:09

Lets look at how many files were copied and the contents of the JSON metadata.

``` r
copied_files <- dir(temp_dir)
copied_files
```

    ## [1] "all_meta.json" "file1.json"    "file1.raw"     "file2.json"   
    ## [5] "file2.raw"

``` r
meta_json <- jsonlite::fromJSON(file.path(temp_dir, "all_meta.json"), simplifyVector = FALSE)
jsonlite::toJSON(meta_json, auto_unbox = TRUE, pretty = TRUE)
```

    ## [
    ##   {
    ##     "file": "file1.raw",
    ##     "saved_path": "/tmp/RtmpF3p4GH/copyfiles-test-122ba73a7b51/file1.raw",
    ##     "original_path": [
    ##       "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file1.raw",
    ##       "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file3.raw"
    ##     ],
    ##     "md5": "8d32638742fe7daad39765205aa5120a"
    ##   },
    ##   {
    ##     "file": "file2.raw",
    ##     "saved_path": "/tmp/RtmpF3p4GH/copyfiles-test-122ba73a7b51/file2.raw",
    ##     "original_path": "/software/R_libs/R340_bioc35/waitcopy/extdata/set1/file2.raw",
    ##     "md5": "dbd76b5cc6d105c8fe077c30887c1389"
    ##   }
    ## ]

### Alternatives (not run)

#### Default

If you just want to copy between 8pm and 6am everyday for however long it will take:

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"))
```

#### Change Start or End Time in Hours

What if the best time to copy files was from 10am until 1pm (13:00)??

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"),
start_time = hours(10), stop_time = hours(13))
```

#### Don't Set a Time Limit

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"),
time_limit = FALSE)
```

#### Stop Checking The Time

If you want the function to give up after trying to check the time, then change the `n_check` variable. If you wanted to stop checking after 3 tries:

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"),
n_check = 3)
```

#### Use Different Renaming Function

An alternative way to handle nasty file names would be to use the `make.names` function:

``` r
wait_copy(file_list, temp_dir, json_meta = file.path(temp_dir, "all_meta.json"),
clean_file_fun = make.names)
```

Note that this is only applied to the `basename` of the file path, i.e. the actual file-name after removing the path in front of the file-name.


NEWS.md
# waitcopy 1.1.0

* Ability to save the meta-data file while waiting for the next iteration of copy events.

* `create_master_meta_data` function can create a full meta-data file from the
set of individual meta-data files in sub-directories.


run_tests.R
# running `Rscript -e "devtools::test()"` does NOT return
# an error code that will make gitlab-ci say the run failed
# R CMD check will, but then you have all the noise from
# check that may not be useful in the early stages of a package
# or analysis. This provides a method to capture that tests
# failed, and notify the CI runner to give the right status.
#
# capture results of running the test

#devtools::install()
library(methods)
test_res <- as.data.frame(devtools::test())

if (sum(test_res$error) > 0) {
  stop_code <- 1
} else if (sum(test_res$failed) > 0) {
  stop_code <- 1
} else if (sum(test_res$warning) > 0) {
  stop_code <- 1
} else {
  stop_code <- 0
}

q('no', stop_code, FALSE)


tests/testthat.R
library(testthat)
library(waitcopy)

test_check("waitcopy")


tests/testthat/test-copying.R
context("copying")

# setup things in a random directory
create_in_temp <- function(dir_loc, create_it = TRUE) {
  temp_path <- tempfile(pattern = paste0("copyfiles-test-", dir_loc))
  if (create_it) {
    dir.create(temp_path)
  }
  temp_path
}
erase <- function(path) unlink(path, recursive = TRUE)

data("iris")

source_dir <- file.path(rprojroot::find_root("DESCRIPTION"), "tests", "testthat")

test_that("copying with duplicates works", {
  target_dir <- create_in_temp("target", FALSE)
  tmp_dir <- create_in_temp("temp")
  on.exit({
    erase(file.path(target_dir))
    erase(file.path(tmp_dir))
  })

  all_files <- dir(file.path(source_dir, c("set1", "set2")), pattern = "raw", full.names = TRUE)

  wait_copy(all_files, target_dir, json_meta = file.path(target_dir, "all_meta_data.json"),
                     tmp_loc = tmp_dir, time_limit = FALSE, pause_file = 0)

  json_metadata <- jsonlite::fromJSON(file.path(target_dir, "all_meta_data.json"), simplifyVector = FALSE)

  # 5 files, but one was a duplicate by md5
  expect_length(json_metadata, 4)

  expect_equal_to_reference(basename(json_metadata[[3]]$saved_path), "file2_dup")

  expect_length(json_metadata[[1]]$original_path, 2)


  new_files <- dir(file.path(source_dir, "set3"), pattern = "raw", full.names = TRUE)
  wait_copy(new_files, target_dir, json_meta = file.path(target_dir, "all_meta_data.json"),
                     tmp_loc = tmp_dir, time_limit = FALSE, pause_file = 0)

  meta_files <- dir(target_dir, pattern = "all_meta_data", full.names = TRUE)

  json_metadata2 <- jsonlite::fromJSON(file.path(target_dir, "all_meta_data.json"), simplifyVector = FALSE)
  expect_length(meta_files, 2)
  expect_length(json_metadata2, 6)
  expect_length(json_metadata2[[1]]$original_path, 3)

})

test_that("timings work", {
  target_dir2 <- create_in_temp("target2")
  tmp_dir2 <- create_in_temp("temp2")
  on.exit({
    erase(file.path(target_dir2))
    erase(file.path(tmp_dir2))
  })

  curr_time <- waitcopy:::get_now_in_local()
  curr_today <- waitcopy:::get_today_in_local()

  now_minus_today <- difftime(curr_time, curr_today, units = "s")
  beg_time <- seconds(now_minus_today + 20)
  end_time <- seconds(now_minus_today + 3600)

  expect_message(wait_copy(file.path(source_dir, "set1", "file1.raw"),
            target_dir2, json_meta = file.path(target_dir2, "all_meta_data.json"), tmp_loc = tmp_dir2,
            start_time = beg_time, stop_time = end_time,
            wait_check = 8),
            "Not allowed to copy yet")

  expect_true(file.exists(file.path(target_dir2, "file1.raw")))
  file_copy_date <- file.mtime(file.path(target_dir2, "file1.raw"))
  expect_true(difftime(file_copy_date, waitcopy:::get_today_in_local() + beg_time) > 0)

  now_minus_today <- difftime(curr_time, curr_today, units = "s")
  beg_time <- seconds(now_minus_today + 50)
  end_time <- seconds(now_minus_today + 3600)

  expect_message(wait_copy(file.path(source_dir, "set1", "file2.raw"),
                           target_dir2, json_meta = file.path(target_dir2, "all_meta_data.json"),
                           tmp_loc = tmp_dir2,
                           start_time = beg_time, stop_time = end_time,
                           wait_check = 2, n_check = 2),
                 "Reached maximum number of wait periods")

  now_minus_today <- difftime(curr_time, curr_today, units = "s")
  beg_time <- seconds(now_minus_today)
  end_time <- seconds(now_minus_today + 3600)

  all_files <- dir(file.path(source_dir, c("set1", "set2")), pattern = "raw", full.names = TRUE)[2:3]

  expect_message(wait_copy(all_files,
                           target_dir2, json_meta = file.path(target_dir2, "all_meta_data.json"),
                           tmp_loc = tmp_dir2,
                           start_time = beg_time, stop_time = end_time,
                           wait_files = 1, pause_wait = 2),
                 "Waiting between sets of files")
})

test_that("warnings appear", {
  check_files <- dir(file.path(source_dir, c("set1", "set2")), full.names = TRUE)
  expect_silent(check_files_exist(check_files))

  check_files <- c(check_files, file.path(source_dir, "idontexist.raw"))
  expect_warning(check_files_exist(check_files), "file list do not exist!")

})


test_that("copy before waiting", {
  skip_on_cran()
  file.create("target_dir3/dummy")
  system("Rscript monitor_files.R -d target_dir3 -l log_file.txt -t 40", intern = FALSE, wait = FALSE)
  Sys.sleep(4)
  has_log_file <- file.exists("log_file.txt")

  skip_if_not(has_log_file)
  #dir.create("target_dir3")
  tmp_dir3 <- create_in_temp("temp3")

  on.exit({
    erase(grep("dummy", dir(file.path("target_dir3"), full.names = TRUE), invert = TRUE, value = TRUE))
    erase(file.path(tmp_dir3))
    erase("log_file.txt")
  })

  curr_time <- waitcopy:::get_now_in_local()
  curr_time2 <- Sys.time()
  curr_today <- waitcopy:::get_today_in_local()

  now_minus_today <- difftime(curr_time, curr_today, units = "s")
  beg_time <- seconds(now_minus_today + 4)
  end_time <- seconds(now_minus_today + 10)

  all_files <- dir(file.path(source_dir, c("set1", "set2")), pattern = "raw", full.names = TRUE)[2:3]

  wait_copy(all_files,
            "target_dir3", json_meta = file.path("target_dir3", "all_meta_data.json"), tmp_loc = tmp_dir3,
            start_time = beg_time, stop_time = end_time,
            wait_check = 4, n_check = 2,
            wait_files = 1, pause_wait = 10)

  # browser(expr = TRUE)

  log_data <- read.table("log_file.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

  log_data <- dplyr::filter(log_data, grepl("target_dir3/all_meta_data.json", file))
  #log_data$check_time <- NULL

  mtimes <- unique(log_data$mtime)

  #browser(expr = TRUE)
  Sys.sleep(20)
  has_2_mtime <- length(mtimes) == 2
  skip_if_not(has_2_mtime)
  expect_equal(length(mtimes), 2)
  expect_gt(as.numeric(difftime(mtimes[1], curr_time2, units = "s")), 14)
  expect_gt(as.numeric(difftime(mtimes[2], curr_time2, units = "s")), 18)


})


R/waitcopy-package.r
#' waitcopy.
#'
#' Copy files during particular times of day and with metadata.
#'
#' @section Functionality:
#'
#' `waitcopy` provides the function `wait_copy` that:
#' - will only copy files during a set time interval
#' - will wait a specific amount of time between file copies
#' - creates a json file with some limited meta-data about the file
#' - removes special characters from the file name
#'
#' @section Motivation:
#'
#' Imagine someone you work with has a hard drive that you need data from, but that
#' hard drive is only accessible via the network, mounted via SAMBA, and there are
#' potentially duplicate files with the same name, files with the same base name
#' that are different, and the file path provides some meta-information about the
#' sample. In addition, the file names have odd characters in them (spaces, colons, etc)
#' that make them a pain to work with from the command line on Linux, so you'd prefer
#' if they weren't there.
#'
#' The files are small, so even copying over the network is fast, but if you copy
#' too many too quickly during the day, you'll get complaints about hitting this shared
#' resource too often by the people who are local to it.
#'
#' @section Solution:
#'
#' So ideally, you want to copy the files only during certain hours, wait a little
#' bit between each copy operation, check for duplicates (via names and md5 hashing),
#' strip the file name of special characters, and note where the file originated.
#'
#' @name waitcopy
#' @docType package
NULL


data-raw/create_testdata.R
source_dir <- file.path(rprojroot::find_root("DESCRIPTION"), "tests", "testthat")

# create some data to work with
dir.create(file.path(source_dir, "set1"))
# two completely different files
dput(iris[, -1], file = file.path(source_dir, "set1", "file1.raw"))
dput(iris[, -3], file = file.path(source_dir, "set1", "file2.raw"))

# and a duplicate in the same directory
dput(iris[, -1], file = file.path(source_dir, "set1", "file3.raw")) # duplicate of file1

dir.create(file.path(source_dir, "set2"))

dput(iris[-10, ], file = file.path(source_dir, "set2", "file4.raw"))
dput(iris[-3, ], file = file.path(source_dir, "set2", "file2.raw"))

# add creating backup and then copy a couple more files
dir.create(file.path(source_dir, "set3"))
dput(iris[, -1], file = file.path(source_dir, "set3", "file5.raw"))
dput(iris[-120, ], file = file.path(source_dir, "set3", "file6.raw"))
dput(iris[, -2], file = file.path(source_dir, "set3", "file7.raw"))


R/copy_files.R
#' clean a file name
#'
#' given a file to copy, generates a new file name stripped of special
#' characters
#'
#' @param in_file the file to copy
#' @param replace_special what to use to replace anything besides dot
#'
#' @details This function creates a new file name that is stripped of special
#' characters. The following characters are replaced by a period: " ", ":", "~",
#' "'". A leading "~" will be replaced by space.
#'
#' @export
#'
#' @return character
#'
#' @examples
#' clean_filename("badly:formatted~file name(sample1)")
#'
clean_filename <- function(in_file, replace_special = "-"){
  use_file <- basename(in_file)

  file_name <- basename(use_file)
  file_name <- gsub(" ", replace_special, file_name, fixed = TRUE)
  file_name <- gsub(":", replace_special, file_name, fixed = TRUE)
  file_name <- gsub("^~", "", file_name)
  file_name <- gsub("~", replace_special, file_name, fixed = TRUE)
  file_name <- gsub("'", replace_special, file_name, fixed = TRUE)
  file_name <- gsub("(", replace_special, file_name, fixed = TRUE)
  file_name <- gsub(")", replace_special, file_name, fixed = TRUE)

  file_name
}

#' copy file
#'
#' Copy a single file from one location to another
#'
#' @param from_file the location of the original file
#' @param to_dir the location to copy the file to
#' @param json_data the json meta information data
#' @param tmp_loc temp location if you want to specify it
#' @param clean_file_fun function used to rename the file
#'
#' @details We want to keep track of information about the copied files, so this
#' function does some stuff to help us out. It strips special characters from
#' the base file name, copies the file to a temp location, calculates the MD5 hash
#' of the file, checks the JSON meta file for matches to the MD5, and if there
#' are none, copies the renamed file to the copy location.
#'
#' If a matching instance of MD5 hashes are found, then the file path is added
#' to the entry for that file location.
#'
#' If a matching file name is found but with a different MD5 hash, then the first
#' 8 characters of the MD5 hash are appended to the file, and it is added to the
#' database.
#'
#' @import digest
#' @importFrom tools file_ext
#' @importFrom purrr map_lgl
#'
#' @export
#' @return list
#'
copy_file <- function(from_file = NULL, to_dir = ".", json_data = NULL, tmp_loc = "/tmp",
                      clean_file_fun = clean_filename){
  stopifnot(!is.null(from_file))

  to_dir <- normalizePath(to_dir)


  base_file <- basename(from_file)
  if (!is.null(clean_file_fun)) {
    base_out <- clean_file_fun(base_file)
  }


  tmp_file <- file.path(tmp_loc, base_out)

  did_copy <- file.copy(from_file, tmp_file)

  if (did_copy) {
    add_file <- TRUE
    md5 <- digest(tmp_file, algo = "md5", file = TRUE)
    if (!is.null(json_data)) {
      match_md5 <- map_lgl(json_data, function(x){md5 %in% x$md5})
      match_file <- map_lgl(json_data, function(x){base_file %in% basename(unlist(x$original_path))})
    } else {
      match_md5 <- FALSE
      match_file <- FALSE
    }

    if (any(match_md5)) {
      #browser(expr = TRUE)
      tmp_json <- json_data[[which(match_md5)]]

      # figure out where the individual json file should be saved
      raw_path <- tmp_json$saved_path
      json_path <- replace_file_extension(raw_path, ".json")

      tmp_json[["original_path"]] <- c(tmp_json[["original_path"]], from_file)

      save_json(tmp_json, json_path)

      json_data[[which(match_md5)]] <- tmp_json

      add_file <- FALSE
    } else if (any(match_file)) {
      fileext <- tools::file_ext(base_out)
      fileext_regex <- paste0(".", fileext, "$")
      base2 <- gsub(fileext_regex, "", base_out)
      base_out <- paste0(base2, "-", substr(md5, 1, 8), ".", fileext)
    }

    if (add_file) {
      file_loc <- file.path(to_dir, base_out)
      did_copy2 <- file.copy(tmp_file, file_loc)
      if (did_copy2) {
        file_data <- vector("list", 1)
        file_data[[1]] <- list(
          file = base_out,
          saved_path = file_loc,
          original_path = from_file,
          md5 = md5
        )
        json_path <- replace_file_extension(file_loc, ".json")
        save_json(file_data, json_path)
        if (is.null(json_data)) {
          json_data <- file_data
        } else {
          json_data <- c(json_data, file_data)
        }
      }

    }
    unlink(tmp_file)
  }

  json_data

}

#' convert to another format
#'
#' given a file path, replace a files extension with a new one
#'
#' @param in_file the file name to work with
#' @param out_extension the new file extension
#'
#' @noRd
replace_file_extension <- function(in_file, out_extension){

  fileext <- tools::file_ext(in_file)
  fileext_regex <- paste0(".", fileext, "$")
  new_file <- gsub(fileext_regex, out_extension, in_file)
  new_file
}

#' save json
#'
#' given a list, saves it to json in a nice way
#'
#' @param list_data the data in list format
#' @param save_loc the location to save it to
#'
#' @noRd
save_json <- function(list_data, save_loc){
  cat(jsonlite::toJSON(list_data, pretty = TRUE, auto_unbox = TRUE), file = save_loc, append = FALSE)
}

#' wait copy
#'
#' Copy files from one location to another, during set hours if desired. This
#' is very useful for copying from networked drives that get a lot of activity
#' during the day.
#'
#' @param file_list a character vector of files to copy from
#' @param to_dir where to copy the files to
#' @param json_meta the json meta flat file
#' @param tmp_loc a temp file location
#' @param clean_file_fun function to use for cleaning up the file name?
#' @param time_limit only copy during a certain time?
#' @param start_time when to start copying
#' @param stop_time when to stop copying
#' @param time_zone what time zone are we in
#' @param wait_check how long to wait before checking again
#' @param n_check how many times to try before giving up
#' @param wait_files how many files before pausing
#' @param pause_wait how long to pause when the wait limit is reached
#' @param pause_file how long to pause between every file
#'
#' @details
#' 1. **Limiting by time of day**: if `time_limit = TRUE`, the `start_time` and
#'   `stop_time` are assumed to be on a
#'   per day basis, so they should be encoded as the number of hours from midnight.
#'   The function actually does a periodic check as to whether the
#'   `start_time` is ahead of it, and if it is not, then it will create
#'   a new time interval for the copying to be allowed. Default is from 8pm (20:00)
#'   to 6am (30:00). The `wait_check` parameter sets how often to wait before checking
#'   the time again (default is 1800 seconds / 30 minutes), and `n_check` parameter
#'   defines how many times to check if the copying can be done (defaults to infinite).
#' 1. **Time Zone**: Provide your time zone so that the time functionality works
#'   properly!
#' 1. **Waiting Between Copies**: In addition to only copying between certain hours,
#'   it is possible to set how long to pause between each file using `pause_file`,
#'   default is 2 seconds, and also a longer interval after copying several files
#'   using `wait_files` (10 files) and `pause_wait` (10 seconds).
#' 1. **Checking for files**: It is recommended before running `wait_copy` to
#'   first run `check_files_exist` on the file list to copy to make sure that
#'   you are passing valid file paths.
#'
#'
#' @import lubridate
#' @importFrom jsonlite fromJSON
#' @importFrom purrr map_chr map
#'
#' @return logical
#' @export
#'
#' @examples
#' \dontrun{
#'
#'   # assume files are in /home/tmp/
#'   # assume current working directory is where to copy to
#'   file_list <- dir(".", full.names = TRUE)
#'   wait_copy(file_list) # copy between 8pm and 6am
#'
#'   # no time limit for copying
#'   wait_copy(file_list, time_limit = FALSE)
#'
#'   # copy from 10am to 1pm (13:00)
#'   wait_copy(file_list, start_time = hours(10), stop_time = hours(13))
#'
#'   # stop checking if can copy after particular number of checks (3)
#'   wait_copy(file_list, n_check = 3)
#'
#'   # check every 30 seconds instead of 30 minutes
#'   wait_copy(file_list, wait_check = 30)
#'
#'   # pause 4 seconds between each file
#'   wait_copy(file_list, pause_file = 4)
#'
#'   # pause 30 seconds after every 20 files
#'   wait_copy(file_list, wait_files = 20, pause_wait = 30)
#'
#'   # don't rename the files
#'   wait_copy(file_list, clean_file_fun = NULL)
#'
#'   # use make.names instead
#'   wait_copy(file_list, clean_file_fun = make.names)
#'
#' }
wait_copy <- function(file_list, to_dir = ".",
                     json_meta = "all_meta_data.json",
                     tmp_loc = "/tmp",
                     clean_file_fun = clean_filename,
                     time_limit = TRUE,
                     start_time = hours(20), stop_time = hours(30),
                     time_zone = NULL, wait_check = 1800, n_check = Inf,
                     wait_files = 10, pause_wait = 10,
                     pause_file = 2){

  if (!dir.exists(to_dir)) {
    dir.create(to_dir)
  }

  if (file.exists(json_meta)) {
    backup_name <- gsub(".json", paste0("-", gsub(" ", "-", as.character(Sys.time())), ".json"), json_meta)
    file.copy(json_meta, backup_name)
    json_data <- jsonlite::fromJSON(json_meta, simplifyVector = FALSE)
  } else {
    json_data <- NULL
  }

  json_md5 <- digest::digest(json_data, "md5")

  # check if we've copied some before, and if so we want to remove them so
  # we don't waste time copying them again.
  if (!(length(json_data) == 0)) {
    #browser()
    previous_files <- unlist(purrr::map(json_data, function(x){x$original_path}))

    file_list <- file_list[!(file_list %in% previous_files)]
  }


  if (time_limit) {
    t_start <- get_today_in_local() + start_time
    t_stop <- get_today_in_local() + stop_time
  } else {
    t_start <- get_today_in_local() - days(10)
    t_stop <- get_today_in_local() + days(10)
  }

  if (is.null(time_zone)) {
    time_zone <- get_tz(now())
  }
  allowed_copy_time <- interval(t_start, t_stop, tz(t_stop))

  # check that the top level directory we are copying from and to
  # actually exists so that we don't try to copy from an invalid location


  to_copy <- length(file_list)
  check_wait_counter <- 0
  i_check <- 0
  did_copy <- 1
  while ((to_copy > 0) && (did_copy <= to_copy)) {
    now1 <- get_now_in_local()
    now2 <- get_now_in_local()
    tmp_int <- interval(now1, now2, tz(t_stop))
    can_copy <- int_overlaps(tmp_int, allowed_copy_time)

    if (can_copy) {
      # copy if we're below the limit for waiting for a time,
      # otherwise, wait for a time. Used to keep us in check on the server
      # and wait some time after copying a reasonable number of files.
      # Note that we also pause a little bit of time after each copy.
      if (check_wait_counter < wait_files) {
        json_data <- copy_file(file_list[did_copy], to_dir, json_data, tmp_loc, clean_file_fun = clean_filename)

        Sys.sleep(pause_file)
        check_wait_counter <- check_wait_counter + 1
        did_copy <- did_copy + 1

      } else {
        Sys.sleep(pause_wait)
        message("Waiting between sets of files! ....")
        check_wait_counter <- 0
      }

    } else {
      # before waiting, we want to check if the meta-data has changed
      # so we can write it before we sit for a long time again
      if (digest::digest(json_data, "md5") != json_md5) {
        save_json(json_data, json_meta)
        json_md5 <- digest::digest(json_data, "md5")
      }
      # if we're not allowed, wait some time before trying again.
      message(paste0("Not allowed to copy yet, waiting! .... ", Sys.time()))
      i_check <- i_check + 1
      if (i_check <= n_check) {
        Sys.sleep(wait_check)
      } else {
        message("Reached maximum number of wait periods, exiting ....")
        break()
      }

    }
  }
  save_json(json_data, json_meta)
}

get_tz <- function(in_time){
  new_time <- as.POSIXlt(in_time)
  new_time$zone
}

get_today_in_local <- function(){
  use_tz <- get_tz(now())
  curr_day <- today() + seconds(1)
  tz(curr_day) <- use_tz
  curr_day
}

get_now_in_local <- function(){
  use_tz <- get_tz(now())
  right_now <- now()
  tz(right_now) <- use_tz
  right_now
}

#' check files
#'
#' check that requested files to copy actually exist, and warn the user
#' if any of them fail.
#'
#' @param file_list the list of files
#' @param n_check how many to check that they exist?
#'
#' @export
check_files_exist <- function(file_list, n_check = 10){
  n_file <- length(file_list)

  if (n_check >= n_file) {
    n_check <- n_file
  }

  files_to_check <- sample(file_list, n_check)

  does_exist <- file.exists(files_to_check)

  if (sum(does_exist) != n_check){
    warning(paste0(sum(!does_exist), " of ", n_check, " files in your file list do not exist!"))
  }
}

#' create meta-data
#'
#' Given a set of directories to search for JSON meta-data files, remake
#' a complete meta-data file and save it to the location specified.
#'
#' @param file_dirs the directories to search
#' @param json_meta where to save the new file
#' @param recursive should they be searched recursively?
#'
#' @export
#'
#' @return list of meta-data, invisibly
create_master_meta_data <- function(file_dirs = ".", json_meta = NULL,
                                    recursive = TRUE){
  all_json <- dir(normalizePath(file_dirs), recursive = recursive, full.names = TRUE, pattern = "json$")
  json_data <- lapply(all_json, function(x){
    #print(x)
    import_json(x)
  })

  #json_lengths <- vapply(json_data, length, numeric(1))

  if (file.exists(json_meta)) {
    backup_name <- gsub(".json", paste0("-", gsub(" ", "-", as.character(Sys.time())), ".json"), json_meta)
    file.copy(json_meta, backup_name)
  }

  if (!is.null(json_meta)) {
    save_json(json_data, json_meta)
    message(paste0("Meta-data saved to ", json_meta))
  }
  invisible(json_data)
}

#' import json
#'
#' import json from a file correctly given some things where things get written
#' differently
#'
#' @param json_file the json file to read
#'
#' @export
import_json <- function(json_file){
  json_data <- jsonlite::fromJSON(json_file, simplifyVector = FALSE)
  if (length(json_data) == 1) {
    out_list <- json_data[[1]]
  } else {
    out_list <- json_data
  }
  out_list
}
